File: v2/src/main/java/com/api/v2/people/domain/PersonAuditTrail.java
package com.api.v2.people.domain;

import org.bson.codecs.pojo.annotations.BsonId;
import org.bson.types.ObjectId;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDate;
import java.time.ZoneId;

@Document
public record PersonAuditTrail(
        @BsonId
        ObjectId id,
        Person person,
        LocalDate createdAt,
        ZoneId createdAtZoneId
) {

    public static PersonAuditTrail create(Person person) {
        return new PersonAuditTrail(
                new ObjectId(),
                person,
                LocalDate.now(),
                ZoneId.systemDefault()
        );
    }
}

File: v2/src/main/kotlin/com/api/v2/people/domain/PersonAuditTrail.kt
package com.api.v2.people.domain

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime
import java.util.UUID

@Document
data class PersonAuditTrail(
    @Id
    val id: String,
    val person: Person,
    val createdAt: LocalDateTime
) {

    companion object {
        fun of( person: Person): PersonAuditTrail {
            return PersonAuditTrail(
                UUID.randomUUID().toString(),
                person,
                LocalDateTime.now()
            )
        }
    }
}
